File: src/main/java/dev/slne/protect/bukkit/region/ProtectionRegion.java
package dev.slne.protect.bukkit.region;

import com.sk89q.worldedit.bukkit.BukkitAdapter;
import com.sk89q.worldedit.math.BlockVector2;
import com.sk89q.worldedit.math.Vector3;
import com.sk89q.worldedit.regions.Region;
import com.sk89q.worldguard.protection.flags.Flags;
import com.sk89q.worldguard.protection.flags.StateFlag;
import com.sk89q.worldguard.protection.managers.RegionManager;
import com.sk89q.worldguard.protection.regions.ProtectedPolygonalRegion;
import com.sk89q.worldguard.protection.regions.ProtectedRegion;
import com.sk89q.worldguard.protection.util.WorldEditRegionConverter;
import dev.slne.protect.bukkit.BukkitMain;
import dev.slne.protect.bukkit.gui.protection.flags.ProtectionFlagsMap;
import dev.slne.protect.bukkit.math.Mth;
import dev.slne.protect.bukkit.math.Mth.EffectiveCostResult;
import dev.slne.protect.bukkit.message.MessageManager;
import dev.slne.protect.bukkit.region.flags.ProtectionFlagsRegistry;
import dev.slne.protect.bukkit.region.info.RegionCreationState;
import dev.slne.protect.bukkit.region.info.RegionInfo;
import dev.slne.protect.bukkit.region.settings.ProtectionSettings;
import dev.slne.protect.bukkit.region.transaction.ProtectionBuyData;
import dev.slne.protect.bukkit.region.visual.Marker;
import dev.slne.protect.bukkit.region.visual.QuickHull;
import dev.slne.protect.bukkit.region.visual.Trail;
import dev.slne.protect.bukkit.user.ProtectionUser;
import dev.slne.transaction.api.TransactionApi;
import dev.slne.transaction.api.currency.Currency;
import dev.slne.transaction.api.transaction.result.TransactionAddResult;
import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Optional;
import net.kyori.adventure.text.logger.slf4j.ComponentLogger;
import org.apache.commons.lang3.RandomStringUtils;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.World;
import org.bukkit.block.Block;
import org.bukkit.block.data.BlockData;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;

public class ProtectionRegion {

    private static final ComponentLogger LOGGER = ComponentLogger.logger("ProtectionRegion");

    private final Location startLocation;
    private final ProtectedRegion expandingProtection;
    private final ProtectionUser protectionUser;
    private final List<Marker> markers;
    private final HashSet<Trail> trails;
    private final ItemStack[] startingInventoryContent;
    private TemporaryProtectionRegion temporaryRegion;
    private List<Marker> boundingMarkers;

    private double worldBorderSize = ProtectionSettings.MAX_DISTANCE_FROM_PROTECTION_START;

    private boolean isProcessingTransaction = false;

    public ProtectionRegion(ProtectionUser protectionUser, ProtectedRegion expandingProtection) {
        this.protectionUser = protectionUser;
        this.expandingProtection = expandingProtection;
        this.startLocation = protectionUser.getBukkitPlayer().getLocation();
        this.startingInventoryContent = protectionUser.getBukkitPlayer().getInventory().getContents().clone();

        this.boundingMarkers = new ArrayList<>();
        this.markers = new ArrayList<>();
        this.trails = new HashSet<>();
    }

    /**
     * Sets the old markers for the expanding region
     */
    public void setExpandingMarkers() {
        if (this.isExpandingRegion()) {
            World world = protectionUser.getBukkitPlayer().getWorld();
            List<BlockVector2> vector2s = expandingProtection.getPoints();

            for (BlockVector2 vector : vector2s) {
                Location location = new Location(world, vector.getX(), 0, vector.getZ());
                location.setY(world.getHighestBlockYAt(location, ProtectionSettings.PROTECTION_HEIGHTMAP) + (double) 1);

                createMarker(location.getBlock(), location.getBlock().getBlockData(), true);
            }
        }
    }

    /**
     * Returns if the protection is expanding
     *
     * @return if the protection is expanding
     */
    public boolean isExpandingRegion() {
        return expandingProtection != null;
    }

    /**
     * Called when a new marker is created
     *
     * @param block        the block of the location
     * @param previousData the previous data in that location
     * @param isExpanding  if the markers get placed because the plugin is expanding the region
     * @return the new marker or null, if marker could not be created
     */
    public Marker createMarker(Block block, BlockData previousData, boolean isExpanding) {
        Marker marker = new Marker(this, block.getLocation(), previousData);
        calculateBoundingMarkers(marker);

        if (!boundingMarkers.contains(marker)) {
            return null;
        }

        // Perform actual state operation
        if (!isExpanding) {
            RegionCreationState state = offerAccepting(false);

            if (state.equals(RegionCreationState.OVERLAPPING)) {
                this.removeMarker(marker);
                this.handleTrails();
                this.calculateBoundingMarkers(marker);

                return null;
            }
        }

        markers.add(marker);
        handleTrails();

        return marker;
    }

    /**
     * Calculates the bounding markers for the given marker
     *
     * @param marker the marker
     */
    private void calculateBoundingMarkers(Marker marker) {
        ArrayList<Marker> newMarkers = new ArrayList<>(this.markers);

        if (marker != null) {
            newMarkers.add(marker);
        }

        boundingMarkers = new QuickHull().quickHull(newMarkers);
    }

    /**
     * Accepts the protection
     *
     * @return the {@link RegionCreationState}
     */
    protected RegionCreationState offerAccepting(boolean calculatePrice) {
        if (boundingMarkers.size() < ProtectionSettings.MIN_MARKERS) {
            protectionUser.sendMessage(MessageManager.getMoreMarkersComponent(boundingMarkers.size()));
            temporaryRegion = null;
            return RegionCreationState.MORE_MARKERS_NEEDED;
        }

        List<BlockVector2> vectors = new ArrayList<>();
        for (Marker marker : boundingMarkers) {
            vectors.add(BlockVector2.at(marker.getLocation().getX(), marker.getLocation().getZ()));
        }

        Player player = this.protectionUser.getBukkitPlayer();

        RegionManager manager = ProtectionUtils.getRegionManager(player.getWorld());
        ProtectedRegion region;

        if (this.isExpandingRegion()) {
            region =
                    new ProtectedPolygonalRegion(expandingProtection.getId(), vectors, ProtectionSettings.MIN_Y_WORLD,
                            ProtectionSettings.MAX_Y_WORLD);
            region.copyFrom(expandingProtection);
        } else {
            String name = player.getName() + "-" + RandomStringUtils.randomAlphabetic(5).toUpperCase();
            region =
                    new ProtectedPolygonalRegion(name, vectors, ProtectionSettings.MIN_Y_WORLD,
                            ProtectionSettings.MAX_Y_WORLD);
            region.getOwners().addPlayer(protectionUser.getLocalPlayer());

            for (ProtectionFlagsMap flagsMap : ProtectionFlagsMap.values()) {
                region.setFlag(flagsMap.getFlag(), flagsMap.getInitialState());
            }

            HashSet<String> owners = new HashSet<>();
            owners.add(protectionUser.getLocalPlayer().getUniqueId().toString());
            region.setFlag(Flags.NONPLAYER_PROTECTION_DOMAINS, owners);
        }

        // Get the center of the region and set the teleport location
        Region worldeditRegion = WorldEditRegionConverter.convertToRegion(region);
        Vector3 center = worldeditRegion.getCenter();
        Location teleportLocation = new Location(player.getWorld(), center.getX(), center.getY(), center.getZ());

        // Set TELE_LOC flag to the center of the region
        region.setFlag(Flags.TELE_LOC, BukkitAdapter.adapt(teleportLocation));

        // Set SURF_PROTECT_FLAG if it does not exist already
        new RegionInfo(player.getWorld(), region);

        // Set SURF_PROTECTION flag to ALLOW
        region.setFlag(ProtectionFlagsRegistry.SURF_PROTECTION, StateFlag.State.ALLOW);

        this.temporaryRegion = new TemporaryProtectionRegion(player.getWorld(), region, manager);
        long area = temporaryRegion.getArea();

        if (this.isExpandingRegion()) {
            temporaryRegion.setEffectiveArea(area - ProtectionUtils.getArea(expandingProtection));
        } else {
            temporaryRegion.setEffectiveArea(area);
        }

        if (temporaryRegion.overlaps(expandingProtection)) {
            protectionUser.sendMessage(MessageManager.getOverlappingRegionsComponent());
            return RegionCreationState.OVERLAPPING;
        } else if (area <= ProtectionSettings.AREA_MIN_BLOCKS) {
            protectionUser.sendMessage(MessageManager.getAreaTooSmallComponent());
            return RegionCreationState.TOO_SMALL;
        } else if (area > ProtectionSettings.AREA_MAX_BLOCKS) {
            protectionUser.sendMessage(MessageManager.getAreaTooBigComponent());
            return RegionCreationState.TOO_LARGE;
        }

        Optional<Currency> currency = TransactionApi.getCurrency("CastCoin");

        if (currency.isEmpty()) {
            protectionUser.sendMessage(MessageManager.getNoCurrencyComponent());
            return RegionCreationState.NO_CURRENCY;
        }

        final EffectiveCostResult effectiveCostResult = Mth.calculateEffectiveCost(teleportLocation, temporaryRegion);
        final double distanceToSpawn = teleportLocation.distance(teleportLocation.getWorld().getSpawnLocation());

        if (effectiveCostResult.effectiveCost() <= 0) {
            protectionUser.sendMessage(MessageManager.getAreaTooSmallComponent());
            return RegionCreationState.TOO_SMALL;
        }

        MessageManager.sendAreaBuyableComponent(protectionUser, area, effectiveCostResult.effectiveCost(), currency.get(), effectiveCostResult.pricePerBlock(), distanceToSpawn);

        this.setTemporaryRegion(temporaryRegion);
        return RegionCreationState.SUCCESS;
    }

    /**
     * Handles removal of marker
     *
     * @param marker the marker
     */
    public void removeMarker(Marker marker) {
        removeMarker0(marker);
        calculateBoundingMarkers(null);
        handleTrails();

        offerAccepting(false);
    }

    /**
     * Handles trails for the markers
     */
    public void handleTrails() {
        HashSet<Trail> newTrails = new HashSet<>();

        for (int i = 0; i < boundingMarkers.size() - 1; i++) {
            Marker marker = boundingMarkers.get(i);
            Marker next = boundingMarkers.get(i + 1);

            Trail trail = new Trail(marker, next, this, true);
            newTrails.add(trail);

            if (!trails.contains(trail)) {
                trails.add(trail);
                trail.start();
            }
        }

        // Add last trail
        if (boundingMarkers.size() >= ProtectionSettings.MIN_MARKERS_LAST_CONNECTION) {
            Marker first = boundingMarkers.get(0);
            Marker last = boundingMarkers.get(boundingMarkers.size() - 1);

            Trail trail = new Trail(first, last, this, true);
            newTrails.add(trail);

            if (!trails.contains(trail)) {
                trails.add(trail);
                trail.start();
            }
        }

        trails.removeIf(trail -> {
            if (!newTrails.contains(trail)) {
                trail.stopTask();

                return true;
            }

            return false;
        });
    }

    /**
     * Calculates the price for the given region
     *
     * @param region        the region
     * @param pricePerBlock the price per block
     * @return the price
     */
    protected double calculateProtectionPrice(TemporaryProtectionRegion region, double pricePerBlock) {
        return region.getEffectiveArea() * pricePerBlock;
    }

    /**
     * Sets the temporary region creation
     *
     * @param temporaryRegion the temporary region creation
     */
    public void setTemporaryRegion(TemporaryProtectionRegion temporaryRegion) {
        this.temporaryRegion = temporaryRegion;
    }

    /**
     * Actually removes the marker
     *
     * @param marker the {@link Marker}
     */
    private void removeMarker0(Marker marker) {
        if (marker.hasPreviousData()) {
            Block block = marker.getLocation().getBlock();
            marker.restorePreviousData();

            if (block.getState().hasMetadata(ProtectionSettings.MARKER_KEY)) {
                block.getState().removeMetadata(ProtectionSettings.MARKER_KEY, BukkitMain.getInstance());
            }
        }

        markers.remove(marker);
    }

    /**
     * Finished the protection
     */
    public void finishProtection() {
        if (temporaryRegion == null) {
            offerAccepting(true);
            return;
        }

        if (temporaryRegion.overlapsUnownedRegion(protectionUser.getLocalPlayer())) {
            protectionUser.sendMessage(MessageManager.getOverlappingRegionsComponent());
            return;
        }


        com.sk89q.worldedit.util.Location worldeditLocation = temporaryRegion.getRegion().getFlag(Flags.TELE_LOC);

        if (worldeditLocation == null) {
            protectionUser.sendMessage(MessageManager.getNoTeleportLocationComponent());
            return;
        }

        Location teleportLocation = BukkitAdapter.adapt(worldeditLocation);
        double pricePerBlock = ProtectionUtils.getProtectionPricePerBlock(teleportLocation);

        double effectiveCost = this.calculateProtectionPrice(temporaryRegion, pricePerBlock);
        BigDecimal effectiveCostBigDecimal = BigDecimal.valueOf(-effectiveCost);

        Optional<Currency> currency = TransactionApi.getCurrency("CastCoin");

        if (currency.isEmpty()) {
            protectionUser.sendMessage(MessageManager.getNoCurrencyComponent());
            return;
        }

        if (isProcessingTransaction) {
            protectionUser.sendMessage(MessageManager.getProtectionAlreadyProcessingComponent());
            return;
        }

        isProcessingTransaction = true;

        protectionUser.hasEnoughCurrency(BigDecimal.valueOf(effectiveCost), currency.get()).thenAcceptAsync(hasEnoughCurrency -> {
            if (!hasEnoughCurrency) {
                protectionUser.sendMessage(MessageManager.getTooExpensiveToBuyComponent());
                isProcessingTransaction = false;
            } else {
                protectionUser.addTransaction(
                        null,
                        effectiveCostBigDecimal,
                        currency.get(),
                        new ProtectionBuyData(this.startLocation != null ? this.startLocation.getWorld() : null, this.temporaryRegion.getRegion())
                ).thenAcceptAsync(transactionAddResult -> {
                    if (transactionAddResult != null && transactionAddResult.equals(TransactionAddResult.SUCCESS)) {
                        this.temporaryRegion.protect();

                        this.removeAllMarkers();
                        Bukkit.getScheduler().runTask(BukkitMain.getInstance(), protectionUser::resetRegionCreation);
                        protectionUser.sendMessage(MessageManager.getProtectionCreatedComponent());
                    } else {
                        protectionUser.sendMessage(MessageManager.getTooExpensiveToBuyComponent());
                    }

                    isProcessingTransaction = false;
                }).exceptionally(throwable -> {
                    LOGGER.error("Error while buying protection", throwable);
                    return null;
                });
            }

        }).exceptionally(throwable -> {
            LOGGER.error("Error while checking if user has enough currency", throwable);
            return null;
        });
    }

    /**
     * Removes all markers
     */
    public void removeAllMarkers() {
        Bukkit.getScheduler().runTask(BukkitMain.getInstance(), () -> {
            for (Marker marker : new ArrayList<>(markers)) {
                removeMarker0(marker);
            }

            calculateBoundingMarkers(null);
            handleTrails();
        });
    }

    /**
     * Cancel the protection
     */
    public void cancelProtection() {
        this.removeAllMarkers();

        protectionUser.sendMessage(MessageManager.getProtectionCanceledComponent());
        protectionUser.resetRegionCreation();
    }

    /**
     * Returns the initial content of the inventory when the region creation is
     * entered
     *
     * @return the initial {@link ItemStack} array
     */
    public ItemStack[] getStartingInventoryContent() {
        return startingInventoryContent;
    }

    /**
     * Returns the marker count left
     *
     * @return the marker count left
     */
    public int getMarkerCountLeft() {
        return getMaxMarkerCount() - getCurrentMarkerCount();
    }

    /**
     * Returns the max marker count
     *
     * @return the max marker count
     */
    public int getMaxMarkerCount() {
        int expandingPoints = 0;

        if (expandingProtection != null) {
            expandingPoints = expandingProtection.getPoints().size();
        }

        return ProtectionSettings.MARKERS + expandingPoints;
    }

    /**
     * Returns the current marker count
     *
     * @return the current marker count
     */
    public int getCurrentMarkerCount() {
        return this.markers.size();
    }

    /**
     * Returns the expanding protection
     *
     * @return the expanding protection
     */
    public ProtectedRegion getExpandingProtection() {
        return expandingProtection;
    }

    /**
     * Returns the start location
     *
     * @return the {@link Location}
     */
    public Location getStartLocation() {
        return startLocation;
    }

    /**
     * Gets the {@link ProtectionUser} for this region
     *
     * @return the {@link ProtectionUser}
     */
    public ProtectionUser getProtectionUser() {
        return protectionUser;
    }

    /**
     * Returns the worldBorderSize
     *
     * @return the worldBorderSize
     */
    public double getWorldBorderSize() {
        return worldBorderSize;
    }

    /**
     * Sets the worldBorderSize
     *
     * @param worldBorderSize the worldBorderSize
     */
    public void setWorldBorderSize(double worldBorderSize) {
        this.worldBorderSize = worldBorderSize;
    }
}

File: src/main/kotlin/dev/slne/surf/protect/paper/region/ProtectionRegion.kt
@file:Suppress("UnstableApiUsage")

package dev.slne.surf.protect.paper.region

import com.github.shynixn.mccoroutine.folia.globalRegionDispatcher
import com.github.shynixn.mccoroutine.folia.launch
import com.sk89q.worldedit.math.BlockVector2
import com.sk89q.worldguard.protection.flags.Flags
import com.sk89q.worldguard.protection.flags.RegionGroup
import com.sk89q.worldguard.protection.flags.StateFlag
import com.sk89q.worldguard.protection.regions.ProtectedPolygonalRegion
import com.sk89q.worldguard.protection.regions.ProtectedRegion
import dev.slne.surf.api.core.messages.adventure.buildText
import dev.slne.surf.api.core.messages.adventure.sendText
import dev.slne.surf.api.core.util.*
import dev.slne.surf.api.paper.util.getHighestBlockYAtBlockCoordinates
import dev.slne.surf.api.paper.util.getXFromChunkKey
import dev.slne.surf.api.paper.util.getZFromChunkKey
import dev.slne.surf.protect.paper.config
import dev.slne.surf.protect.paper.event.ProtectionCreateEvent
import dev.slne.surf.protect.paper.math.Mth
import dev.slne.surf.protect.paper.message.Messages
import dev.slne.surf.protect.paper.permission.ProtectPermissionRegistry
import dev.slne.surf.protect.paper.plugin
import dev.slne.surf.protect.paper.region.flags.EditableProtectionFlags
import dev.slne.surf.protect.paper.region.flags.ProtectionFlagsRegistry
import dev.slne.surf.protect.paper.region.info.RegionCreationState
import dev.slne.surf.protect.paper.region.info.RegionInfo
import dev.slne.surf.protect.paper.region.settings.ProtectionSettings
import dev.slne.surf.protect.paper.region.visual.Marker
import dev.slne.surf.protect.paper.region.visual.QuickHull
import dev.slne.surf.protect.paper.region.visual.Trail
import dev.slne.surf.protect.paper.region.visual.visualizer.ProtectionVisualizerManager
import dev.slne.surf.protect.paper.user.ProtectionUser
import dev.slne.surf.protect.paper.util.*
import io.papermc.paper.math.BlockPosition
import it.unimi.dsi.fastutil.objects.ObjectLinkedOpenHashSet
import it.unimi.dsi.fastutil.objects.ObjectList
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.future.await
import kotlinx.coroutines.withContext
import org.apache.commons.lang3.RandomStringUtils
import org.bukkit.Chunk
import org.bukkit.ChunkSnapshot
import org.bukkit.block.data.BlockData
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import java.lang.ref.WeakReference
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.math.min
import kotlin.math.roundToInt

class ProtectionRegion(
    val protectionUser: ProtectionUser,
    val player: Player,
    playerInventoryContent: Array<ItemStack?>,
    val expandingProtection: ProtectedRegion? = null
) {
    val startLocation = player.location
    val startingInventoryContent = playerInventoryContent
    var worldBorderSize = config.protection.maxDistanceFromStart

    private val markers = mutableObjectListOf<Marker>(config.markers.amount)
    private val trails = mutableObjectSetOf<Trail>(config.markers.amount)
    private var hull = ObjectLinkedOpenHashSet<Marker>()

    private val tmpBuffer = mutableObjectListOf<Marker>()
    private var tempRegion: TempProtectionRegion? = null

    private val isProcessingTransaction = AtomicBoolean(false)

    val markerCountLeft: Int get() = maxMarkerCount - currentMarkerCount
    val maxMarkerCount: Int
        get() = config.markers.amount + (expandingProtection?.points?.size ?: 0)
    val currentMarkerCount: Int get() = markers.size

    suspend fun setCornerMarkers() {
        val region = expandingProtection ?: return
        val world = player.world
        val points = region.points

        val byChunk = mutableLong2ObjectMapOf<ObjectList<BlockVector2>>(points.size / 4 + 1)
        for (point in points) {
            val namespacedKey = Chunk.getChunkKey(point.x() shr 4, point.z() shr 4)
            val list = byChunk.computeIfAbsent(namespacedKey) { mutableObjectListOf() }
            list.add(point)
        }

        val snapshots = coroutineScope {
            byChunk.keys.mapAsync { namespacedKey ->
                namespacedKey to world.getChunkAtAsync(
                    getXFromChunkKey(namespacedKey),
                    getZFromChunkKey(namespacedKey)
                )
                    .await()
                    .getChunkSnapshot(true, false, false, false)
            }.toMap(mutableLong2ObjectMapOf<ChunkSnapshot>(byChunk.size))
        }

        val it = byChunk.long2ObjectEntrySet().fastIterator()
        while (it.hasNext()) {
            val entry = it.next()
            val namespacedKey = entry.longKey
            val pointsInChunk = entry.value
            val snapshot = snapshots[namespacedKey]
                ?: error("ChunkSnapshot for namespacedKey $namespacedKey not found")
            for (point in pointsInChunk) {
                val x = point.x()
                val z = point.z()
                val y =
                    min(snapshot.getHighestBlockYAtBlockCoordinates(x, z) + 1, world.maxHeight - 1)
                val pos = point.toBlockPosition(y)
                val data = snapshot.getBlockDataAt(pos.blockX(), pos.blockY(), pos.blockZ())
                createMarker(pos, data, isExpanding = true)
            }
        }
    }

    fun createMarker(pos: BlockPosition, previousData: BlockData, isExpanding: Boolean): Marker? {
        val candidate = Marker(WeakReference(player.world), this, pos, previousData)
        if (!updateHullPreview(candidate)) return null

        // Check overlap state when not expanding
        if (!isExpanding && offerAccepting() == RegionCreationState.OVERLAPPING) {
            handleTrails()
            restoreHull()
            return null
        }

        markers.add(candidate)
        plugin.launch { candidate.place() }
        handleTrails()

        return candidate
    }

    private fun updateHullPreview(candidate: Marker): Boolean {
        tmpBuffer.clear()
        tmpBuffer.addAll(markers)
        tmpBuffer.add(candidate)
        val preview = QuickHull.compute(tmpBuffer)
        return if (preview.contains(candidate)) {
            hull.clear()
            hull.addAll(preview)
            true
        } else false
    }

    private fun restoreHull() {
        if (markers.size < 3) {
            hull.clear()
            hull.addAll(markers)
            return
        }
        hull.clear()
        hull.addAll(QuickHull.compute(markers))
    }

    /**
     * Accepts the protection
     *
     * @return the [RegionCreationState]
     */
    private fun getDiscountFactor(): Double {
        return if (player.hasPermission(ProtectPermissionRegistry.PROTECTION_DISCOUNT)) {
            config.pricing.discountModifier
        } else {
            1.0
        }
    }

    private fun offerAccepting(): RegionCreationState {
        if (hull.size < config.markers.minAmount) {
            protectionUser.sendMessage(Messages.Protecting.moreMarkers(hull.size))
            tempRegion = null
            return RegionCreationState.MORE_MARKERS_NEEDED
        }

        val vectors = mutableObjectListOf<BlockVector2>()
        for (marker in hull) {
            vectors.add(marker.toBlockVector2())
        }

        val world = player.world
        val manager = world.getRegionManager()
        val region: ProtectedRegion

        if (expandingProtection != null) {
            region = ProtectedPolygonalRegion(
                expandingProtection.id,
                vectors,
                world.minHeight,
                world.maxHeight - 1
            )
            region.copyFrom(expandingProtection)
        } else {
            val name = player.name + "-" + RandomStringUtils.secureStrong()
                .nextAlphabetic(ProtectionSettings.RANDOM_NAME_LENGTH)
                .uppercase()

            region = ProtectedPolygonalRegion(
                name,
                vectors,
                world.minHeight,
                world.maxHeight - 1
            )
            region.owners.addPlayer(protectionUser.localPlayer)

            for (flagsMap in EditableProtectionFlags.entries) {
                if (flagsMap.isPlayerRelated) {
                    region.setFlag(flagsMap.flag, StateFlag.State.ALLOW)
                    if ((flagsMap.initialState ?: StateFlag.State.ALLOW) == StateFlag.State.DENY) {
                        region.setFlag(flagsMap.flag.regionGroupFlag, RegionGroup.MEMBERS)
                    } else {
                        region.setFlag(flagsMap.flag.regionGroupFlag, null)
                    }
                } else {
                    region.setFlag(flagsMap.flag, flagsMap.initialState)
                }
            }

            region.setFlag(
                Flags.NONPLAYER_PROTECTION_DOMAINS,
                objectSetOf(player.uniqueId.toString())
            )
        }

        // Get the center of the region and set the teleport location
        val center = region.fastCenter().toVector3()
        val centerLoc = com.sk89q.worldedit.util.Location(
            world.toWorldEdit(),
            center.x(),
            center.y(),
            center.z()
        )

        region.setFlag(Flags.TELE_LOC, centerLoc)

        // Set SURF_PROTECT_FLAG if it does not exist already
        RegionInfo(region)

        // Set SURF_PROTECTION flag to ALLOW
        region.setFlag(ProtectionFlagsRegistry.SURF_PROTECTION, StateFlag.State.ALLOW)


        val tmpRegion = TempProtectionRegion(world, region, manager).also { tempRegion = it }
        val tmpVolume = tmpRegion.volume
        if (expandingProtection != null) {
            tmpRegion.effectiveVolume = tmpVolume - expandingProtection.fixedVolume()
        }

        return when {
            tmpRegion.overlapsUnownedRegion(protectionUser.localPlayer) -> RegionCreationState.OVERLAPPING.also {
                protectionUser.sendMessage(Messages.Protecting.overlappingRegions)
            }

            tmpVolume <= config.area.minBlocks -> RegionCreationState.TOO_SMALL.also {
                protectionUser.sendMessage(Messages.Protecting.areaTooSmall)
            }

            tmpVolume > config.area.maxBlocks -> RegionCreationState.TOO_LARGE.also {
                protectionUser.sendMessage(Messages.Protecting.areaTooBig)
            }

            else -> {
                val currency = config.currency
                val discountFactor = getDiscountFactor()
                val (effectiveCost, pricePerBlock, spawnDistance) = Mth.calculateEffectiveCost(
                    centerLoc,
                    tmpRegion,
                    discountFactor
                )

                if (effectiveCost <= 0) {
                    protectionUser.sendMessage(Messages.Protecting.areaTooSmall)
                    RegionCreationState.TOO_SMALL
                } else {
                    if (spawnDistance < config.pricing.spawnProtectionPerBlock) {
                        protectionUser.sendMessage(buildText {
                            appendErrorPrefix()
                            error("Das Grundstück liegt zu nah am Spawn.")
                        })
                        return RegionCreationState.TOO_NEAR_FROM_SPAWN
                    }

                    protectionUser.sendMessage(
                        Messages.Protecting.offer(
                            tmpVolume,
                            effectiveCost,
                            currency.currency,
                            pricePerBlock,
                            spawnDistance,
                            discountFactor
                        )
                    )
                    RegionCreationState.SUCCESS
                }
            }
        }
    }

    /**
     * Handles removal of marker
     *
     * @param marker the marker
     */
    suspend fun removeMarker(marker: Marker) {
        marker.restorePreviousData()
        markers.remove(marker)
        restoreHull()
        handleTrails()
        offerAccepting()
    }

    /**
     * Handles trails for the markers
     */
    fun handleTrails() {
        val newTrails = mutableObjectSetOf<Trail>()
        val hullSeq = hull.toObjectList()
        val size = hullSeq.size
        if (size < 2) {
            trails.forEach { it.close() }
            trails.clear()
            return
        }

        fun ensureTrail(a: Marker, b: Marker) {
            val trail = Trail(a, b, this, expandingProtection == null)
            trails.add(trail)
            newTrails.add(trail)
        }

        for (i in 0 until size - 1) {
            ensureTrail(hullSeq[i], hullSeq[i + 1])
        }

        // Add last trail
        if (size >= ProtectionSettings.MIN_MARKERS_LAST_CONNECTION) {
            ensureTrail(hullSeq[size - 1], hullSeq[0])
        }

        // Remove obsolete trails
        trails.removeIf { trail ->
            if (!newTrails.contains(trail)) {
                trail.close()
                true
            } else {
                trail.start()
                false
            }
        }
    }

    suspend fun finishProtection() {
        val tempRegion = tempRegion ?: return run {
            offerAccepting()
        }

        if (tempRegion.overlapsUnownedRegion(protectionUser.localPlayer)) {
            protectionUser.sendMessage(Messages.Protecting.overlappingRegions)
            return
        }

        if (tempRegion.volume <= config.area.minBlocks) {
            protectionUser.sendMessage(Messages.Protecting.areaTooSmall)
            return
        }

        val centerLoc = tempRegion.region.getFlag(Flags.TELE_LOC) ?: return run {
            protectionUser.sendMessage(Messages.Protecting.noTpPointFound)
        }

        val discountFactor = getDiscountFactor()
        val costResult = Mth.calculateEffectiveCost(centerLoc, tempRegion, discountFactor)
        val cost = costResult.effectiveCost.roundToInt()
        val costBD = (-cost).toBigDecimal()
        val currency = config.currency.currency

        if (costResult.pricePerBlock == Double.MAX_VALUE) {
            protectionUser.sendMessage(buildText {
                appendErrorPrefix()
                error("Das Grundstück liegt zu nah am Spawn.")
            })
            return
        }

        if (!isProcessingTransaction.compareAndSet(false, true)) {
            protectionUser.sendMessage(Messages.Protecting.alreadyProcessingTransaction)
            return
        }

        try {
            if (protectionUser.transactionUser.balance(currency) < costBD.abs()) {
                protectionUser.sendMessage(Messages.Protecting.tooExpensiveToBuy)
                return
            }

            val result = protectionUser.transactionUser.withdraw(costBD, currency, false)

            if (result.success) {
                tempRegion.protect()
                removeAllMarkers()
                protectionUser.resetRegionCreation(false)
                protectionUser.bukkitPlayer?.sendText {
                    appendSuccessPrefix()
                    success("Das Grundstück wurde erstellt.")
                }

                if (expandingProtection != null) {
                    ProtectionVisualizerManager.onRegionCornerChange(tempRegion.region)
                } else {
                    ProtectionVisualizerManager.onRegionCreated(
                        startLocation.world,
                        tempRegion.region
                    )

                    applyDefaultFlags(tempRegion.region)
                }

                withContext(plugin.globalRegionDispatcher) {
                    protectionUser.bukkitPlayer?.let {
                        ProtectionCreateEvent(it).callEvent()
                    }
                }
            } else {
                protectionUser.sendMessage(Messages.Protecting.tooExpensiveToBuy)
            }

        } finally {
            isProcessingTransaction.set(false)
        }
    }

    private fun applyDefaultFlags(region: ProtectedRegion) {
        for (flag in EditableProtectionFlags.entries) {

            val state = flag.initialState ?: StateFlag.State.ALLOW

            if (flag.isPlayerRelated) {
                region.setFlag(flag.flag, StateFlag.State.ALLOW)
                if (state == StateFlag.State.DENY) {
                    region.setFlag(flag.flag.regionGroupFlag, RegionGroup.MEMBERS)
                } else {
                    region.setFlag(flag.flag.regionGroupFlag, null)
                }
                continue
            }

            region.setFlag(flag.flag, state)
        }
    }

    /**
     * Removes all markers
     */
    suspend fun removeAllMarkers(shutdown: Boolean = false) {
        coroutineScope {
            markers.toObjectList().map { marker ->
                async {
                    marker.restorePreviousData(shutdown)
                }
            }.awaitAll()
        }
        markers.clear()
        restoreHull()
        handleTrails()
    }

    /**
     * Cancel the protection
     */
    suspend fun cancelProtection(shutdown: Boolean = false) {
        removeAllMarkers(shutdown)

        protectionUser.bukkitPlayer?.sendText {
            appendInfoPrefix()
            info("Der Grundstückserstellungsprozess wurde abgebrochen.")
        }

        protectionUser.resetRegionCreation(true, shutdown)
    }
}